{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2d3d09d3-2460-424d-b9bd-f31de7514015",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘SPATAwrappers’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:SPATA2’:\n",
      "\n",
      "    hlpr_normalize_imap, hlpr_smooth_shiny\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.1     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32m%||%()\u001b[39m   masks \u001b[34mSPATAwrappers\u001b[39m::%||%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "library(SPATA2)\n",
    "library(SPATAwrappers)\n",
    "library(tidyverse)\n",
    "# library(future)\n",
    "# plan(multisession , workers = 4)\n",
    "# options(future.globals.maxSize= 160*1024^3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "709611e1-f069-47ba-92d6-f6eef192b14c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.3 (2024-02-29)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: CentOS Linux 8 (Core)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /data/person/lab/fengcz/conda/envs/spata2/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=C.UTF-8    LC_NUMERIC=C        LC_TIME=C          \n",
       " [4] LC_COLLATE=C        LC_MONETARY=C       LC_MESSAGES=C      \n",
       " [7] LC_PAPER=C          LC_NAME=C           LC_ADDRESS=C       \n",
       "[10] LC_TELEPHONE=C      LC_MEASUREMENT=C    LC_IDENTIFICATION=C\n",
       "\n",
       "time zone: Asia/Shanghai\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] lubridate_1.9.3          forcats_1.0.0            stringr_1.5.1           \n",
       " [4] dplyr_1.1.4              purrr_1.0.2              readr_2.1.5             \n",
       " [7] tidyr_1.3.1              tibble_3.2.1             ggplot2_3.5.1           \n",
       "[10] tidyverse_2.0.0          SPATAwrappers_0.0.0.9000 SPATA2_2.0.4            \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RColorBrewer_1.1-3          jsonlite_1.8.8             \n",
       "  [3] magrittr_2.0.3              spatstat.utils_3.0-5       \n",
       "  [5] zlibbioc_1.48.0             vctrs_0.6.5                \n",
       "  [7] ROCR_1.0-11                 spatstat.explore_3.2-6     \n",
       "  [9] RCurl_1.98-1.16             base64enc_0.1-3            \n",
       " [11] htmltools_0.5.8.1           S4Arrays_1.2.0             \n",
       " [13] SparseArray_1.2.2           sctransform_0.4.1          \n",
       " [15] parallelly_1.37.1           KernSmooth_2.23-24         \n",
       " [17] htmlwidgets_1.6.4           ica_1.0-3                  \n",
       " [19] plyr_1.8.9                  plotly_4.10.4              \n",
       " [21] zoo_1.8-12                  uuid_1.1-1                 \n",
       " [23] igraph_2.0.3                mime_0.12                  \n",
       " [25] lifecycle_1.0.4             pkgconfig_2.0.3            \n",
       " [27] Matrix_1.6-5                R6_2.5.1                   \n",
       " [29] fastmap_1.2.0               GenomeInfoDbData_1.2.11    \n",
       " [31] MatrixGenerics_1.14.0       fitdistrplus_1.1-11        \n",
       " [33] future_1.33.2               shiny_1.8.1.1              \n",
       " [35] digest_0.6.36               colorspace_2.1-0           \n",
       " [37] patchwork_1.2.0             S4Vectors_0.40.2           \n",
       " [39] tensor_1.5                  Seurat_4.4.0               \n",
       " [41] irlba_2.3.5.1               GenomicRanges_1.54.1       \n",
       " [43] progressr_0.14.0            timechange_0.3.0           \n",
       " [45] fansi_1.0.6                 spatstat.sparse_3.1-0      \n",
       " [47] polyclip_1.10-6             httr_1.4.7                 \n",
       " [49] abind_1.4-5                 compiler_4.3.3             \n",
       " [51] withr_3.0.0                 tiff_0.1-12                \n",
       " [53] MASS_7.3-60.0.1             DelayedArray_0.28.0        \n",
       " [55] tools_4.3.3                 units_0.8-5                \n",
       " [57] lmtest_0.9-40               httpuv_1.6.15              \n",
       " [59] future.apply_1.11.2         goftest_1.2-3              \n",
       " [61] glue_1.7.0                  EBImage_4.44.0             \n",
       " [63] nlme_3.1-165                promises_1.3.0             \n",
       " [65] grid_4.3.3                  anndata_0.7.5.6            \n",
       " [67] pbdZMQ_0.3-10               Rtsne_0.17                 \n",
       " [69] cluster_2.1.6               reshape2_1.4.4             \n",
       " [71] generics_0.1.3              gtable_0.3.5               \n",
       " [73] spatstat.data_3.1-2         tzdb_0.4.0                 \n",
       " [75] hms_1.1.3                   data.table_1.15.4          \n",
       " [77] sp_2.1-4                    utf8_1.2.4                 \n",
       " [79] XVector_0.42.0              spatstat.geom_3.2-9        \n",
       " [81] BiocGenerics_0.48.1         RcppAnnoy_0.0.22           \n",
       " [83] ggrepel_0.9.5               RANN_2.6.1                 \n",
       " [85] pillar_1.9.0                spam_2.10-0                \n",
       " [87] IRdisplay_1.1               later_1.3.2                \n",
       " [89] splines_4.3.3               lattice_0.22-6             \n",
       " [91] deldir_2.0-4                survival_3.7-0             \n",
       " [93] tidyselect_1.2.1            locfit_1.5-9.10            \n",
       " [95] SingleCellExperiment_1.24.0 miniUI_0.1.1.1             \n",
       " [97] pbapply_1.7-2               gridExtra_2.3              \n",
       " [99] IRanges_2.36.0              SummarizedExperiment_1.32.0\n",
       "[101] scattermore_1.2             stats4_4.3.3               \n",
       "[103] Biobase_2.62.0              SPATAData_0.0.0.9000       \n",
       "[105] matrixStats_1.3.0           fftwtools_0.9-11           \n",
       "[107] stringi_1.8.4               lazyeval_0.2.2             \n",
       "[109] evaluate_0.24.0             codetools_0.2-20           \n",
       "[111] cli_3.6.3                   uwot_0.1.16                \n",
       "[113] IRkernel_1.3.2              xtable_1.8-4               \n",
       "[115] reticulate_1.38.0           repr_1.1.6                 \n",
       "[117] munsell_0.5.1               Rcpp_1.0.12                \n",
       "[119] GenomeInfoDb_1.38.8         spatstat.random_3.2-3      \n",
       "[121] globals_0.16.3              png_0.1-8                  \n",
       "[123] parallel_4.3.3              assertthat_0.2.1           \n",
       "[125] dotCall64_1.1-1             jpeg_0.1-10                \n",
       "[127] bitops_1.0-7                listenv_0.9.1              \n",
       "[129] viridisLite_0.4.2           scales_1.3.0               \n",
       "[131] ggridges_0.5.6              SeuratObject_5.0.2         \n",
       "[133] leiden_0.4.3.1              crayon_1.5.3               \n",
       "[135] rlang_1.1.4                 cowplot_1.1.3              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a4afca2-b022-475c-8333-8e1fccd4a134",
   "metadata": {},
   "source": [
    "# Load SPATA2 infercnv functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4d2d06e6-f0b9-41e4-9e35-af4fd4b7c16f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#' @title Check sample input\n",
    "#'\n",
    "#' @description A member of the \\code{adjusting-check_*()}-family. Takes a character\n",
    "#' vector of sample names and checks which of these exist.\n",
    "#'\n",
    "#' Returns an adjusted sample-vector or raises an error.\n",
    "#' @param object A valid spata-object.\n",
    "#' @param of_sample This argument is currently inactive. It might be reactivated when\n",
    "#' spata-objects can store more than one sample.\n",
    "#' @param desired_length The length the input must have.\n",
    "#'\n",
    "#' @keywords internal\n",
    "check_sample <- function(object,\n",
    "                         of_sample = \"\",\n",
    "                         desired_length = NULL,\n",
    "                         ...){\n",
    "\n",
    "\n",
    "  # 0. Default sample -------------------------------------------------------\n",
    "\n",
    "  # 'of_sample' is currently not in use\n",
    "  if(FALSE){\n",
    "\n",
    "  confuns::is_vec(of_sample, mode = \"character\", ...)\n",
    "\n",
    "  if(base::all(of_sample == \"\")){\n",
    "\n",
    "    of_sample <- getSampleName(object)[1]\n",
    "\n",
    "    if(base::length(getSampleName(object)) > 1){\n",
    "\n",
    "      base::message(glue::glue(\"No sample specified. Defaulting to first sample: '{of_sample}'.\"))\n",
    "\n",
    "    }\n",
    "\n",
    "  }\n",
    "\n",
    "  # 1. Check which samples are in the object --------------------------------\n",
    "\n",
    "  if(!base::is.character(of_sample) | base::length(of_sample) == 0){\n",
    "\n",
    "    stop(\"Please specify the sample with its name as a character vector of length > 0.\")\n",
    "\n",
    "  } else if(base::all(of_sample == \"all\")){\n",
    "\n",
    "    of_sample <- getSampleNames(object = object)\n",
    "\n",
    "    if(!base::is.null(desired_length) && base::length(of_sample) != desired_length){\n",
    "\n",
    "      stop(stringr::str_c(\"Number of samples specified needs to be: \", desired_length, \". \",\n",
    "                          \"Setting 'of_sample' to 'all' results in \",\n",
    "                          base::length(of_sample), \" samples.\", sep = \"\"))\n",
    "\n",
    "    }\n",
    "\n",
    "    return(of_sample)\n",
    "\n",
    "  }  else if(!base::any(of_sample %in% object@samples)){\n",
    "\n",
    "    stop(\"Could not find any of the specified samples in specified object.\")\n",
    "\n",
    "  } else if(base::any(of_sample %in% getSampleNames(object))){\n",
    "\n",
    "    samples_found <- object@samples[object@samples %in% of_sample]\n",
    "\n",
    "    if(base::length(samples_found) != base::length(of_sample)){\n",
    "\n",
    "      not_found <- of_sample[!of_sample %in% object@samples]\n",
    "      n_not_found <- base::length(not_found)\n",
    "\n",
    "      not_found <- stringr::str_c(not_found, collapse = \"', '\")\n",
    "\n",
    "      ref <- base::ifelse(n_not_found > 1, \"samples\", \"sample\")\n",
    "\n",
    "      warning(glue::glue(\"Did not find {n_not_found} {ref}: {not_found}\"))\n",
    "\n",
    "    }\n",
    "\n",
    "  }\n",
    "\n",
    "  # -----\n",
    "\n",
    "  # 2. Check if length of samples found coincides with desired length -------\n",
    "\n",
    "  if(!base::is.null(desired_length) &&\n",
    "     base::length(samples_found) != desired_length){\n",
    "\n",
    "    base::stop(stringr::str_c(\"Number of samples specified needs to be:\", desired_length, sep = \" \"))\n",
    "\n",
    "  }\n",
    "\n",
    "  }\n",
    "\n",
    "  # -----\n",
    "\n",
    "  return(getSampleNames(object))\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "setInitiationInfo <- function(object,\n",
    "                              additional_input = list()){\n",
    "\n",
    "  ce <- rlang::caller_env()\n",
    "\n",
    "  init_fn <- rlang::caller_fn()\n",
    "\n",
    "  init_frame <- base::sys.parent()\n",
    "\n",
    "  init_call <- base::sys.call(which = init_frame)\n",
    "\n",
    "  init_fn_name <- base::as.character(init_call)[1]\n",
    "\n",
    "  init_args <-\n",
    "    rlang::fn_fmls_names(fn = init_fn)\n",
    "\n",
    "  init_args <- init_args[init_args != \"...\"]\n",
    "\n",
    "  init_args_input <-\n",
    "    purrr::map(\n",
    "      .x = init_args,\n",
    "      .f = function(arg){\n",
    "\n",
    "        base::parse(text = arg) %>%\n",
    "          base::eval(envir = ce)\n",
    "\n",
    "      }\n",
    "    ) %>%\n",
    "    purrr::set_names(nm = init_args)\n",
    "\n",
    "  init_args_input <-\n",
    "    init_args_input[!base::names(init_args_input) %in% c(\"cds\",  \"coords_df\", \"count_mtr\", \"expr_mtr\", \"object\", \"seurat_object\")]\n",
    "\n",
    "  init_args_input <-\n",
    "    c(init_args_input, additional_input)\n",
    "\n",
    "  initiation_list <- list(\n",
    "    init_fn = init_fn_name,\n",
    "    input = init_args_input,\n",
    "    time = base::Sys.time()\n",
    "  )\n",
    "\n",
    "  object@information$initiation <-\n",
    "    initiation_list\n",
    "\n",
    "  base::return(object)\n",
    "\n",
    "}\n",
    "\n",
    "initiateSpataObject_modified <- function(coords_df,\n",
    "                                         count_mtr,\n",
    "                                         feature_df = NULL,\n",
    "                                         sample_name,\n",
    "                                         image = NULL,\n",
    "                                         directory_spata = NULL,\n",
    "                                         directory_seurat = NULL,\n",
    "                                         combine_with_wd = FALSE,\n",
    "                                         gene_set_path = NULL,\n",
    "                                         SCTransform = FALSE,\n",
    "                                         NormalizeData = list(normalization.method = \"LogNormalize\", scale.factor = 1000),\n",
    "                                         FindVariableFeatures = list(selection.method = \"vst\", nfeatures = 2000),\n",
    "                                         ScaleData = TRUE,\n",
    "                                         RunPCA = list(npcs = 60),\n",
    "                                         FindNeighbors = list(dims = 1:30),\n",
    "                                         FindClusters = list(resolution = 0.8),\n",
    "                                         RunTSNE = TRUE,\n",
    "                                         RunUMAP = list(dims = 1:30),\n",
    "                                         verbose = TRUE){\n",
    "\n",
    "    # 1. Control --------------------------------------------------------------\n",
    "    confuns::give_feedback(\n",
    "      msg = \"Starting initiation\",\n",
    "      verbose = verbose\n",
    "    )\n",
    "\n",
    "    confuns::is_value(x = sample_name, mode = \"character\", ref = \"sample_name\")\n",
    "\n",
    "    confuns::check_data_frame(\n",
    "      df = coords_df,\n",
    "      var.class = list(\"barcodes\" = \"character\", \"x\" = c(\"double\", \"integer\", \"numeric\"), \"y\" = c(\"double\", \"integer\", \"numeric\")),\n",
    "      ref = \"coords_df\"\n",
    "    )\n",
    "\n",
    "    if(!methods::is(count_mtr, \"Matrix\")){\n",
    "\n",
    "      base::stop(\"'count_mtr'-input needs to be of type 'Matrix'.\")\n",
    "\n",
    "    } else if(base::is.null(base::colnames(count_mtr))){\n",
    "\n",
    "      base::stop(\"'count_mtr'-input needs to have column names\")\n",
    "\n",
    "    } else if(base::is.null(base::rownames(count_mtr))){\n",
    "\n",
    "      base::stop(\"'count_mtr'-input needs to have row names.\")\n",
    "\n",
    "    }\n",
    "\n",
    "    barcodes_count_mtr <- base::colnames(count_mtr) %>% base::sort()\n",
    "    barcodes_coords_df <- dplyr::pull(coords_df, var = \"barcodes\") %>% base::sort()\n",
    "\n",
    "    # check identical barcodes\n",
    "    if(!base::identical(barcodes_count_mtr, barcodes_coords_df)){\n",
    "\n",
    "      base::stop(\"Barcodes of 'coords_df'-input and column names of 'count_mtr'-input need to be identical.\")\n",
    "\n",
    "    }\n",
    "\n",
    "\n",
    "    # -----\n",
    "\n",
    "    # 2. Passing data ---------------------------------------------------------\n",
    "\n",
    "    counts <- count_mtr\n",
    "\n",
    "    seurat_object <-\n",
    "      Seurat::CreateSeuratObject(counts = counts, meta.data = feature_df)\n",
    "\n",
    "    # processed_seurat_object <-\n",
    "    #   process_seurat_object(\n",
    "    #     seurat_object = seurat_object,\n",
    "    #     SCTransform = SCTransform,\n",
    "    #     NormalizeData = NormalizeData,\n",
    "    #     FindVariableFeatures = FindVariableFeatures,\n",
    "    #     ScaleData = ScaleData,\n",
    "    #     RunPCA = RunPCA,\n",
    "    #     FindNeighbors = FindNeighbors,\n",
    "    #     FindClusters = FindClusters,\n",
    "    #     RunTSNE = RunTSNE,\n",
    "    #     RunUMAP = RunUMAP,\n",
    "    #     verbose = verbose\n",
    "    #   )\n",
    "\n",
    "\n",
    "    # 3. Passing features and images ------------------------------------------\n",
    "\n",
    "    spata_object <-\n",
    "      transformSeuratToSpata(\n",
    "        seurat_object = seurat_object,\n",
    "        assay_name = \"RNA\",\n",
    "        sample_name = sample_name,\n",
    "        gene_set_path = gene_set_path,\n",
    "        method = \"single_cell\",\n",
    "        verbose = verbose\n",
    "      )\n",
    "\n",
    "    spata_object <-\n",
    "      setCoordsDf(object = spata_object, coords_df = coords_df) %>%\n",
    "      setImage(object = ., image = image)\n",
    "\n",
    "    spata_object <- setInitiationInfo(spata_object)\n",
    "\n",
    "    # 4. Save and return object -----------------------------------------------\n",
    "\n",
    "    # save seurat\n",
    "    if(base::is.character(directory_seurat)){\n",
    "\n",
    "      spata_object <-\n",
    "        base::tryCatch({\n",
    "\n",
    "          spata_object_return <-\n",
    "            saveCorrespondingSeuratObject(\n",
    "              seurat_object = processed_seurat_object,\n",
    "              object = spata_object,\n",
    "              directory_seurat = directory_seurat,\n",
    "              combine_with_wd = combine_with_wd\n",
    "            )\n",
    "\n",
    "          spata_object_return\n",
    "\n",
    "        }, error = function(error){\n",
    "\n",
    "          base::warning(glue::glue(\"Attempt to save seurat-object under '{directory_seurat}' failed with the following error message: {error}\"))\n",
    "\n",
    "          spata_object\n",
    "\n",
    "        })\n",
    "\n",
    "    }\n",
    "\n",
    "    # save spata\n",
    "    if(base::is.character(directory_spata)){\n",
    "\n",
    "      spata_object <-\n",
    "        base::tryCatch({\n",
    "\n",
    "          spata_object_ret <-\n",
    "            saveSpataObject(\n",
    "              object = spata_object,\n",
    "              directory_spata = directory_spata,\n",
    "              combine_with_wd = combine_with_wd\n",
    "            )\n",
    "\n",
    "          spata_object_ret\n",
    "\n",
    "        }, error = function(error){\n",
    "\n",
    "          base::warning(glue::glue(\"Attempt to save spata-object under '{directory_spata}' failed with the following error message: {error}\"))\n",
    "\n",
    "          spata_object\n",
    "\n",
    "        })\n",
    "\n",
    "    }\n",
    "\n",
    "    confuns::give_feedback(\n",
    "      msg = \"Initiation finished.\",\n",
    "      verbose = verbose\n",
    "    )\n",
    "\n",
    "    # -----\n",
    "\n",
    "    base::return(spata_object)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "check_availability <- function(test, ref_x, ref_fns){\n",
    "\n",
    "  if(!base::isTRUE(test)){\n",
    "\n",
    "    msg <- glue::glue(\"Could not find {ref_x}. Make sure to use {ref_fns}.\")\n",
    "\n",
    "    confuns::give_feedback(msg = msg, fdb.fn = \"stop\", with.time = FALSE)\n",
    "\n",
    "  }\n",
    "\n",
    "  base::invisible(TRUE)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "hlpr_assign_arguments <- function(object){\n",
    "\n",
    "  check_object(object)\n",
    "\n",
    "  default_instructions <- getDefaultInstructions(object)\n",
    "\n",
    "  ce <- rlang::caller_env()\n",
    "\n",
    "  cfn <- rlang::caller_fn()\n",
    "\n",
    "  cargs <- rlang::fn_fmls_names(fn = cfn)\n",
    "\n",
    "  default_args <- cargs[cargs %in% methods::slotNames(default_instructions_object)]\n",
    "\n",
    "  for(arg in default_args){\n",
    "\n",
    "    arg_value <-\n",
    "      base::parse(text = arg) %>%\n",
    "      base::eval(envir = ce)\n",
    "\n",
    "    if(base::is.null(arg_value)){\n",
    "\n",
    "      arg_value <- methods::slot(default_instructions, name = arg)\n",
    "\n",
    "      base::assign(\n",
    "        x = arg,\n",
    "        value = arg_value,\n",
    "        envir = ce\n",
    "      )\n",
    "    }\n",
    "\n",
    "  }\n",
    "\n",
    "  base::invisible(TRUE)\n",
    "\n",
    "}\n",
    "\n",
    "default_colors <- viridis::viridis(15)\n",
    "\n",
    "default_instructions_object <-\n",
    "  methods::new(Class = \"default_instructions\",\n",
    "               average_genes = FALSE,\n",
    "               binwidth = 1,\n",
    "               clrp = \"milo\",\n",
    "               clrsp = \"inferno\",\n",
    "               colors = default_colors,\n",
    "               complete = TRUE,\n",
    "               display_facets = TRUE,\n",
    "               display_image = FALSE,\n",
    "               display_labels = TRUE,\n",
    "               display_legend = TRUE,\n",
    "               display_points = FALSE,\n",
    "               display_residuals = TRUE,\n",
    "               display_title = FALSE,\n",
    "               display_trajectory_parts = FALSE,\n",
    "               max_adj_pval = 0.05,\n",
    "               method_aggl = \"ward.D\",\n",
    "               method_dist = \"euclidean\",\n",
    "               method_de = \"wilcox\",\n",
    "               method_dr = \"umap\",\n",
    "               method_gs = \"mean\",\n",
    "               method_hclust = \"complete\",\n",
    "               method_ovl = \"classic\",\n",
    "               method_padj = \"fdr\",\n",
    "               min_lfc = 0,\n",
    "               normalize = TRUE,\n",
    "               n_highest_lfc = 50,\n",
    "               n_lowest_pval = 50,\n",
    "               n_pcs = 10,\n",
    "               position = \"fill\",\n",
    "               pt_alpha = 0.9,\n",
    "               pt_clr = \"lightgrey\",\n",
    "               pt_clrp = \"milo\",\n",
    "               pt_clrsp = \"inferno\",\n",
    "               pt_fill = \"black\",\n",
    "               pt_shape = 21,\n",
    "               pt_size = 2,\n",
    "               pt_size_fixed = TRUE,\n",
    "               relevel = FALSE,\n",
    "               scales = \"free\",\n",
    "               sgmt_clr = \"black\",\n",
    "               sgmt_size = 2.5,\n",
    "               show_rownames = FALSE,\n",
    "               show_colnames = FALSE,\n",
    "               smooth = TRUE,\n",
    "               smooth_clr = \"red\",\n",
    "               smooth_method = \"loess\",\n",
    "               smooth_se = TRUE,\n",
    "               smooth_span = 0.25,\n",
    "               uniform_genes = \"discard\",\n",
    "               verbose = TRUE)\n",
    "\n",
    "#' @title Set cnv-results\n",
    "#'\n",
    "#' @inherit check_sample params\n",
    "#' @inherit set_dummy details\n",
    "#'\n",
    "#' @param cnv_list The list containing the results from \\code{runCnvAnalysis()}.\n",
    "#'\n",
    "#' @return An updated spata-object.\n",
    "#' @export\n",
    "#'\n",
    "\n",
    "setCnvResults <- function(object, cnv_list, of_sample = NA){\n",
    "\n",
    "  check_object(object)\n",
    "\n",
    "  of_sample <- check_sample(object = object, of_sample = of_sample, of.length = 1)\n",
    "\n",
    "  object@cnv[[of_sample]] <- cnv_list\n",
    "\n",
    "  base::return(object)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "#' @title Obtain copy-number-variations results\n",
    "#'\n",
    "#' @description Provides convenient access to the results of \\code{runCnvAnalysis()}.\n",
    "#'\n",
    "#' @inherit check_sample params\n",
    "#'\n",
    "#' @return A named list.\n",
    "#' @export\n",
    "#'\n",
    "\n",
    "getCnvResults <- function(object, of_sample = NA){\n",
    "\n",
    "  check_object(object)\n",
    "\n",
    "  of_sample <- check_sample(object = object, of_sample = of_sample, of.length = 1)\n",
    "\n",
    "  res_list <- object@cnv[[of_sample]]\n",
    "\n",
    "  check_availability(test = !base::identical(x = res_list, y = list()),\n",
    "                     ref_x = \"CNV results\",\n",
    "                     ref_fns = \"function 'runCnvAnalysis()\")\n",
    "\n",
    "  base::return(res_list)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "#' @title Obtain features names under which cnv-analysis results are stored.\n",
    "#'\n",
    "#' @description Returns a character vector of feature names referring to the\n",
    "#' barcode-spots chromosomal gains and losses as computed by \\code{runCnvAnalysis()}.\n",
    "#'\n",
    "#' @inherit check_sample params\n",
    "#'\n",
    "#' @return Character vector.\n",
    "#' @export\n",
    "#'\n",
    "\n",
    "getCnvFeatureNames <- function(object, of_sample = NA){\n",
    "\n",
    "  check_object(object)\n",
    "\n",
    "  of_sample <- check_sample(object = object, of_sample = of_sample, of.length = 1)\n",
    "\n",
    "  cnv_results <- getCnvResults(object = object, of_sample = of_sample)\n",
    "\n",
    "  prefix <- cnv_results$prefix\n",
    "\n",
    "  chromosomes <-\n",
    "    cnv_results$regions_df %>%\n",
    "    base::rownames() %>%\n",
    "    stringr::str_remove_all(pattern = \"p$|q$\") %>%\n",
    "    base::unique()\n",
    "\n",
    "  cnv_feature_names <- stringr::str_c(prefix, chromosomes)\n",
    "\n",
    "  base::return(cnv_feature_names)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "hlpr_run_cnva_pca <- function(object, n_pcs = 30, of_sample = NA, ...){\n",
    "\n",
    "  check_object(object)\n",
    "\n",
    "  of_sample <- check_sample(object, of_sample = of_sample, desired_length = 1)\n",
    "\n",
    "  cnv_res <- getCnvResults(object, of_sample = of_sample)\n",
    "\n",
    "  cnv_mtr <- cnv_res$cnv_mtr\n",
    "\n",
    "  pca_res <- irlba::prcomp_irlba(x = base::t(cnv_mtr), n = n_pcs, ...)\n",
    "\n",
    "  pca_df <-\n",
    "    base::as.data.frame(x = pca_res[[\"x\"]]) %>%\n",
    "    dplyr::mutate(barcodes = base::colnames(cnv_mtr), sample = {{of_sample}}) %>%\n",
    "    dplyr::select(barcodes, sample, dplyr::everything()) %>%\n",
    "    tibble::as_tibble()\n",
    "\n",
    "  cnv_res$dim_red$pca <- pca_df\n",
    "\n",
    "  object <- setCnvResults(object, cnv_list = cnv_res, of_sample = of_sample)\n",
    "\n",
    "  base::return(object)\n",
    "\n",
    "\n",
    "}\n",
    "\n",
    "runCnvAnalysis_modified <- function(object,\n",
    "                           ref_annotation = cnv_ref[[\"annotation\"]], # data.frame denoting reference data as reference\n",
    "                           ref_mtr = cnv_ref[[\"mtr\"]], # reference data set of healthy tissue\n",
    "                           ref_regions = cnv_ref[[\"regions\"]], # chromosome positions\n",
    "                           gene_pos_df = SPATA2::gene_pos_df,\n",
    "                           directory_cnv_folder = \"data-development/cnv-results\", # output folder\n",
    "                           directory_regions_df = NA, # deprecated (chromosome positions)\n",
    "                           n_pcs = 30,\n",
    "                           cnv_prefix = \"Chr\",\n",
    "                           save_infercnv_object = TRUE,\n",
    "                           verbose = TRUE,\n",
    "                           of_sample = NA,\n",
    "                           CreateInfercnvObject = list(ref_group_names = \"ref\"),\n",
    "                           require_above_min_mean_expr_cutoff = list(min_mean_expr_cutoff = 0.1),\n",
    "                           require_above_min_cells_ref = list(min_cells_per_gene = 3),\n",
    "                           normalize_counts_by_seq_depth = list(),\n",
    "                           anscombe_transform = list(),\n",
    "                           log2xplus1 = list(),\n",
    "                           apply_max_threshold_bounds = list(),\n",
    "                           smooth_by_chromosome = list(window_length = 101, smooth_ends = TRUE),\n",
    "                           center_cell_expr_across_chromosome = list(method = \"median\"),\n",
    "                           subtract_ref_expr_from_obs = list(inv_log = TRUE),\n",
    "                           invert_log2 = list(),\n",
    "                           clear_noise_via_ref_mean_sd = list(sd_amplifier = 1.5),\n",
    "                           remove_outliers_norm = list(),\n",
    "                           define_signif_tumor_subclusters = list(p_val = 0.05, hclust_method = \"ward.D2\", cluster_by_groups = TRUE, partition_method = \"qnorm\"),\n",
    "                           plot_cnv = list(k_obs_groups = 5, cluster_by_groups = TRUE, output_filename = \"infercnv.outliers_removed\", color_safe_pal = FALSE,\n",
    "                                           x.range = \"auto\", x.center = 1, output_format = \"pdf\", title = \"Outliers Removed\")\n",
    "                           ){\n",
    "\n",
    "  # 1. Control --------------------------------------------------------------\n",
    "\n",
    "  hlpr_assign_arguments(object)\n",
    "  of_sample <- check_sample(object = object, of_sample = of_sample, of.lenght = 1)\n",
    "\n",
    "  confuns::are_values(c(\"save_infercnv_object\"), mode = \"logical\")\n",
    "\n",
    "  confuns::check_directories(directories = directory_cnv_folder, type = \"folders\")\n",
    "\n",
    "  if(!base::is.na(directory_regions_df)){\n",
    "\n",
    "    base::message(\n",
    "      \"Redirecting input for argument 'directory_regions_df' (deprecated) to \",\n",
    "      \"argument 'ref_regions'. Please use 'ref_regions' instead.\"\n",
    "    )\n",
    "\n",
    "    ref_regions <- directory_regions_df\n",
    "\n",
    "    base::warning(\n",
    "      \"The argument 'directory_regions_df' is deprecated in favor of 'ref_regions'. \",\n",
    "      \"See documentation for more details.\"\n",
    "      )\n",
    "\n",
    "  }\n",
    "\n",
    "  # -----\n",
    "\n",
    "\n",
    "  # 2. Data extraction ------------------------------------------------------\n",
    "\n",
    "  # preparing object derived data\n",
    "  count_mtr <- getCountMatrix(object = object, of_sample = of_sample)\n",
    "\n",
    "  obj_anno <-\n",
    "    getFeatureDf(object = object, of_sample = of_sample) %>%\n",
    "    dplyr::select(barcodes, sample) %>%\n",
    "    tibble::column_to_rownames(var = \"barcodes\")\n",
    "\n",
    "  # reading and preparing reference data\n",
    "  confuns::give_feedback(msg = \"Checking input for reference data.\", verbose = verbose)\n",
    "\n",
    "  if(base::is.character(ref_mtr) && stringr::str_detect(ref_mtr, pattern = \"\\\\.RDS$\")){\n",
    "\n",
    "    confuns::give_feedback(\n",
    "      msg = glue::glue(\"Reading in reference matrix from directory '{ref_mtr}'.\"),\n",
    "      verbose = verbose\n",
    "      )\n",
    "\n",
    "    ref_mtr <- base::readRDS(file = ref_mtr)\n",
    "\n",
    "    confuns::give_feedback(msg = \"Done.\", verbose = verbose)\n",
    "\n",
    "  } else if(base::is.matrix(ref_mtr)){\n",
    "\n",
    "    ref_mtr <- ref_mtr\n",
    "\n",
    "  } else {\n",
    "\n",
    "    base::stop(\"Input for argument 'ref_mtr' must either be a directory leading to an .RDS-file or a matrix.\")\n",
    "\n",
    "  }\n",
    "\n",
    "\n",
    "  if(base::is.character(ref_annotation) && stringr::str_detect(ref_annotation, pattern = \"\\\\.RDS$\")){\n",
    "\n",
    "    confuns::give_feedback(\n",
    "      msg = glue::glue(\"Reading in reference annotation from directory '{ref_annotation}'.\"),\n",
    "      verbose = verbose\n",
    "    )\n",
    "\n",
    "    ref_anno <- base::readRDS(file = ref_annotation)\n",
    "\n",
    "    confuns::give_feedback(msg = \"Done.\", verbose = verbose)\n",
    "\n",
    "  } else if(base::is.data.frame(ref_annotation)){\n",
    "\n",
    "    ref_anno <- ref_annotation\n",
    "\n",
    "  } else {\n",
    "\n",
    "    base::stop(\"Input for argument 'ref_annotation' must either be a directory leading to an .RDS-file or a data.frame.\")\n",
    "\n",
    "  }\n",
    "\n",
    "\n",
    "  # combine data sets\n",
    "  confuns::give_feedback(msg = \"Combining input and reference data.\", verbose = verbose)\n",
    "\n",
    "  genes_inter <-\n",
    "    base::intersect(x = base::rownames(count_mtr), y = base::rownames(ref_mtr)) %>% base::unique()\n",
    "\n",
    "  if(base::length(genes_inter) < 500){\n",
    "\n",
    "    msg <- \"Less than 500 genes match ref and count matrix.\"\n",
    "\n",
    "    confuns::give_feedback(msg = msg, fdb.fn = \"stop\", with.time = FALSE)\n",
    "\n",
    "  }\n",
    "\n",
    "  expr_inter <- base::cbind(count_mtr[genes_inter, ], ref_mtr[genes_inter, ])\n",
    "\n",
    "  anno_inter <- base::rbind(obj_anno, ref_anno)\n",
    "\n",
    "  # read and process gene positions\n",
    "  confuns::give_feedback(msg = \"Getting gene positions.\", verbose = verbose)\n",
    "\n",
    "  if(base::is.character(ref_regions) && stringr::str_detect(ref_regions, pattern = \"\\\\.RDS$\")){\n",
    "\n",
    "    regions_df <- base::readRDS(file = ref_regions)\n",
    "\n",
    "  } else if(base::is.data.frame(ref_regions)) {\n",
    "\n",
    "    regions_df <- ref_regions\n",
    "\n",
    "  } else {\n",
    "\n",
    "    base::stop(\"Input for argument 'ref_regions' must either be a directory leading to an .RDS-file or a data.frame\")\n",
    "\n",
    "  }\n",
    "\n",
    "  if(base::is.data.frame(gene_pos_df)){\n",
    "\n",
    "    confuns::check_data_frame(\n",
    "      df = gene_pos_df,\n",
    "      var.class = list(\n",
    "        # ensembl_gene_id = \"character\",\n",
    "        hgnc_symbol = \"character\",\n",
    "        chromosome_name = \"character\",\n",
    "        start_position = \"integer\",\n",
    "        end_position = \"integer\"\n",
    "      )\n",
    "    )\n",
    "\n",
    "\n",
    "  } else {\n",
    "\n",
    "    gene_pos_df <-\n",
    "      CONICSmat::getGenePositions(gene_names = base::rownames(expr_inter))\n",
    "\n",
    "  }\n",
    "\n",
    "\n",
    "  # -----\n",
    "\n",
    "\n",
    "  # 3. Analysis pipeline ----------------------------------------------------\n",
    "\n",
    "  gene_order_df <-\n",
    "    dplyr::select(gene_pos_df, chromosome_name, start_position, end_position, hgnc_symbol) %>%\n",
    "    magrittr::set_rownames(value = gene_pos_df$hgnc_symbol)\n",
    "\n",
    "  confuns::give_feedback(msg = \"Starting analysis pipeline.\", verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"CreateInfercnvObject\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"raw_counts_matrix\" = expr_inter,\n",
    "                     \"annotations_file\" = anno_inter,\n",
    "                     \"gene_order_file\" = gene_order_df\n",
    "                     )\n",
    "    )\n",
    "\n",
    "\n",
    "  confuns::give_feedback(msg = glue::glue(\"Removing genes from matrix with mean expression below threshold.\"), verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"require_above_min_mean_expr_cutoff\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "      )\n",
    "\n",
    "  confuns::give_feedback(msg = \"Removing low quality barcode spots.\", verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"require_above_min_cells_ref\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "\n",
    "  confuns::give_feedback(msg = \"Normalizing counts by sequencing depth.\", verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"normalize_counts_by_seq_depth\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "\n",
    "  confuns::give_feedback(msg = \"Conducting anscombe and logarithmic transformation.\", verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"anscombe_transform\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"log2xplus1\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "  confuns::give_feedback(msg = \"Applying maximal threshold bounds.\", verbose = verbose)\n",
    "\n",
    "  # threshold <-\n",
    "  #   base::mean(base::abs(infercnv:::get_average_bounds(infercnv_obj))) %>%\n",
    "  #   base::round(digits = 2)\n",
    "\n",
    "  threshold = 2.6\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"apply_max_threshold_bounds\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj, \"threshold\" = threshold),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "\n",
    "  confuns::give_feedback(msg = \"Smoothing by chromosome.\", verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"smooth_by_chromosome\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "\n",
    "  confuns::give_feedback(msg = \"Centering cell expression across chromosome.\", verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"center_cell_expr_across_chromosome\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "  confuns::give_feedback(msg = \"Subtracting reference expression from observed expression.\", verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"subtract_ref_expr_from_obs\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "  confuns::give_feedback(msg = \"Clearing noise via reference mean standard deviation.\", verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"invert_log2\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "\n",
    "    )\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"clear_noise_via_ref_mean_sd\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "\n",
    "  confuns::give_feedback(msg = \"Removing outliers.\", verbose = verbose)\n",
    "\n",
    "  infercnv_obj <-\n",
    "    confuns::call_flexibly(\n",
    "      fn = \"remove_outliers_norm\",\n",
    "      fn.ns = \"infercnv\",\n",
    "      fn.ns.sep = \":::\",\n",
    "      default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "      v.fail = infercnv_obj\n",
    "    )\n",
    "\n",
    "\n",
    "#   confuns::give_feedback(msg = \"Defining significant tumor subclusters.\", verbose = verbose)\n",
    "\n",
    "#   infercnv_obj <-\n",
    "#     confuns::call_flexibly(\n",
    "#       fn = \"define_signif_tumor_subclusters\",\n",
    "#       fn.ns = \"infercnv\",\n",
    "#       fn.ns.sep = \":::\",\n",
    "#       default = list(\"infercnv_obj\" = infercnv_obj),\n",
    "#       v.fail = infercnv_obj\n",
    "#     )\n",
    "\n",
    "\n",
    "  confuns::give_feedback(msg = \"Copy number variation pipeline completed.\", verbose = verbose)\n",
    "\n",
    "  if(base::isTRUE(save_infercnv_object)){\n",
    "\n",
    "    save_dir <- stringr::str_c(directory_cnv_folder, \"infercnv-obj.RDS\", sep = \"/\")\n",
    "\n",
    "    msg <- glue::glue(\"Saving infercnv-object under '{save_dir}'.\")\n",
    "\n",
    "    confuns::give_feedback(msg = msg, verbose = verbose)\n",
    "\n",
    "    base::saveRDS(infercnv_obj, file = save_dir)\n",
    "\n",
    "  }\n",
    "\n",
    "#   confuns::give_feedback(msg = \"Plotting results.\", verbose = verbose)\n",
    "\n",
    "#   plot_results <-\n",
    "#     confuns::call_flexibly(\n",
    "#       fn = \"plot_cnv\",\n",
    "#       fn.ns = \"infercnv\",\n",
    "#       fn.ns.sep = \":::\",\n",
    "#       default = list(\"infercnv_obj\" = infercnv_obj, \"out_dir\" = directory_cnv_folder),\n",
    "#       v.fail = NULL\n",
    "#     )\n",
    "\n",
    "\n",
    "#   # work around a weird error in plot_cnv()\n",
    "#   if(base::is.null(plot_results)){\n",
    "\n",
    "#     confuns::give_feedback(\n",
    "#       msg = \"infercnv:::plot_cnv() failed. Attempting to plot with default setting.\",\n",
    "#       verbose = TRUE\n",
    "#     )\n",
    "\n",
    "#     plot_results <-\n",
    "#       base::tryCatch({\n",
    "\n",
    "#         infercnv::plot_cnv(infercnv_obj = infercnv_obj, out_dir = directory_cnv_folder)\n",
    "\n",
    "#       }, error = function(error){\n",
    "\n",
    "#         NULL\n",
    "\n",
    "#       })\n",
    "\n",
    "#     if(base::is.null(plot_results)){\n",
    "\n",
    "#       confuns::give_feedback(\n",
    "#         msg = \"inferncnv::plot_cnv() failed with default setting, too.\",\n",
    "#         verbose = TRUE\n",
    "#       )\n",
    "\n",
    "#     }\n",
    "\n",
    "#     if(base::isTRUE(save_infercnv_object)){\n",
    "\n",
    "#       msg <-\n",
    "#         glue::glue(\n",
    "#           \"The infercnv-object has been saved under '{save_dir}'.\",\n",
    "#           \"Please try to plot the heatmap manually.\"\n",
    "#           )\n",
    "\n",
    "#       confuns::give_feedback(msg = msg, verbose = TRUE)\n",
    "\n",
    "#     } else {\n",
    "\n",
    "#       msg <-\n",
    "#         glue::glue(\n",
    "#           \"Please consider to run runCnvAnalysis() again with argument 'save_infercnv_object' set to TRUE.\",\n",
    "#           \"This way you can plot the results manually.\"\n",
    "#         )\n",
    "\n",
    "#       confuns::give_feedback(msg = msg, verbose = TRUE)\n",
    "\n",
    "#     }\n",
    "\n",
    "#     msg <- \"If the error in infercnv:::plot_cnv() persists, consider to open an issue at https://github.com/theMILOlab/SPATA2/issues.\"\n",
    "\n",
    "#     confuns::give_feedback(msg = msg, verbose = TRUE)\n",
    "\n",
    "#   }\n",
    "\n",
    "  # ----\n",
    "\n",
    "\n",
    "  # 4. Storing results ------------------------------------------------------\n",
    "\n",
    "  # result_dir <-\n",
    "  #   stringr::str_c(directory_cnv_folder, \"/\", plot_cnv$output_filename, \".observations.txt\")\n",
    "\n",
    "    \n",
    "  # infercnv_obj = readRDS(paste0(directory_cnv_folder,'/infercnv-obj.RDS'))\n",
    "\n",
    "  # results <- utils::read.table(result_dir)\n",
    "  results = infercnv_obj@expr.data\n",
    "    \n",
    "  barcodes <- base::colnames(results)\n",
    "\n",
    "  confuns::give_feedback(msg = \"Summarizing cnv-results by chromosome.\", verbose = verbose)\n",
    "\n",
    "  # join cnv results (per gene) with chromosome positions and summarize by chromosome\n",
    "  ordered_cnv_df <-\n",
    "    base::as.data.frame(results) %>%\n",
    "    tibble::rownames_to_column(\"hgnc_symbol\") %>%\n",
    "    dplyr::left_join(., gene_pos_df, by = \"hgnc_symbol\") %>%\n",
    "    dplyr::group_by(chromosome_name) %>%\n",
    "    dplyr::select(chromosome_name, dplyr::any_of(barcodes)) %>%\n",
    "    dplyr::summarise_all(base::mean) %>%\n",
    "    dplyr::mutate(Chr = stringr::str_c(cnv_prefix, chromosome_name)) %>%\n",
    "    dplyr::select(Chr, dplyr::everything()) %>%\n",
    "    dplyr::ungroup() %>%\n",
    "    dplyr::select(-chromosome_name)\n",
    "\n",
    "  cnames <- c(\"barcodes\", ordered_cnv_df$Chr)\n",
    "\n",
    "  ordered_cnv_df2 <-\n",
    "    dplyr::select(ordered_cnv_df, -Chr) %>%\n",
    "    base::t() %>%\n",
    "    base::as.data.frame() %>%\n",
    "    tibble::rownames_to_column(var = \"barcodes\") %>%\n",
    "    magrittr::set_colnames(value = cnames) %>%\n",
    "    dplyr::mutate(barcodes = stringr::str_replace_all(string = barcodes, pattern = \"\\\\.\", replacement = \"-\")) %>%\n",
    "    dplyr::mutate(dplyr::across(dplyr::starts_with(match = cnv_prefix), .fns = base::as.numeric)) %>%\n",
    "    tibble::as_tibble()\n",
    "\n",
    "  # add results to spata object\n",
    "  confuns::give_feedback(msg = \"Adding results to the spata-object's feature data.\", verbose = verbose)\n",
    "\n",
    "  # feature variables\n",
    "  object <-\n",
    "    addFeatures(\n",
    "      object = object,\n",
    "      feature_df = ordered_cnv_df2,\n",
    "      overwrite = TRUE\n",
    "      )\n",
    "\n",
    "  # cnv matrix\n",
    "  base::colnames(results) <-\n",
    "    stringr::str_replace_all(\n",
    "      string = base::colnames(results),\n",
    "      pattern = \"\\\\.\",\n",
    "      replacement = \"-\"\n",
    "      )\n",
    "\n",
    "  cnv_mtr <- base::as.matrix(results)\n",
    "\n",
    "  # cnv list\n",
    "  cnv_list <-\n",
    "    list(\n",
    "      prefix = cnv_prefix,\n",
    "      cnv_df = ordered_cnv_df2,\n",
    "      cnv_mtr = cnv_mtr,\n",
    "      gene_pos_df = gene_pos_df,\n",
    "      regions_df = regions_df\n",
    "      )\n",
    "\n",
    "  object <-\n",
    "    setCnvResults(object = object, cnv_list = cnv_list, of_sample = of_sample)\n",
    "\n",
    "#   confuns::give_feedback(msg = \"Computing PCA based on cnv results.\", verbose = verbose)\n",
    "\n",
    "#   object <-\n",
    "#     hlpr_run_cnva_pca(object, n_pcs = n_pcs, of_sample = of_sample)\n",
    "\n",
    "#   # cnv clustering - hierarchical\n",
    "#   cnv_res <- getCnvResults(object, of_sample = of_sample)\n",
    "\n",
    "#   cnv_pca_df <- dplyr::select(cnv_res$dim_red$pca, -sample)\n",
    "\n",
    "#   cnv_hclust <-\n",
    "#     confuns::initiate_hclust_object(\n",
    "#       hclust.data = tibble::column_to_rownames(cnv_pca_df, var = \"barcodes\"),\n",
    "#       key.name = \"barcodes\"\n",
    "#       )\n",
    "\n",
    "#   clustering_list <- list(hierarchical = cnv_hclust)\n",
    "\n",
    "#   cnv_res$clustering <- clustering_list\n",
    "\n",
    "#   object <- setCnvResults(object, cnv_list = cnv_res, of_sample = of_sample)\n",
    "\n",
    "  # -----\n",
    "    \n",
    "    \n",
    "  confuns::give_feedback(msg = \"Infercnv Done.\", verbose = verbose)\n",
    "\n",
    "  base::return(object)\n",
    "\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ee01228-7456-4ac9-9365-842a97ffed77",
   "metadata": {},
   "source": [
    "# Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "98ea4075-4b22-45fd-8dee-d4e3e5235160",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: SingleCellExperiment \n",
       "dim: 16147 38012 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(16147): A1BG A2M ... ZYX ZZEF1\n",
       "rowData names(0):\n",
       "colnames(38012): P0_FT_1_50_7550 P0_FT_1_50_7750 ... P0_FT_1_9150_9150\n",
       "  P0_FT_1_4900_8700\n",
       "colData names(35): n_genes_by_counts log1p_n_genes_by_counts ... myCAF\n",
       "  MMP11- Cell_type_refined\n",
       "reducedDimNames(1): spatial\n",
       "mainExpName: NULL\n",
       "altExpNames(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "P0_FT_1 = readRDS('./P0_FT_1.sce.rds')\n",
    "P0_FT_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b9e165ce-8410-4c0e-93c7-45256d624d2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: SingleCellExperiment \n",
       "dim: 16147 87959 \n",
       "metadata(0):\n",
       "assays(1): counts\n",
       "rownames(16147): A1BG A2M ... ZYX ZZEF1\n",
       "rowData names(0):\n",
       "colnames(87959): P6_YF_1_0_2550 P6_YF_1_0_3250 ... P6_YF_1_7950_15350\n",
       "  P6_YF_1_8050_15400\n",
       "colData names(35): n_genes_by_counts log1p_n_genes_by_counts ... myCAF\n",
       "  MMP11- Cell_type_refined\n",
       "reducedDimNames(1): spatial\n",
       "mainExpName: NULL\n",
       "altExpNames(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "P6_YF_1 = readRDS('./P6_YF_1.sce.rds')\n",
    "P6_YF_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e1b027fb-2f29-4d5e-bad4-dbbb6c89352c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv_ref = SPATA2::cnv_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2c90e6c4-22bc-4e79-a2a1-e30fb4b66da3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pos_df = read.table('gene_pos.txt',sep='\\t')\n",
    "colnames(gene_pos_df) = c('hgnc_symbol','chromosome_name','start_position','end_position')\n",
    "gene_pos_df$chromosome_name = gsub('chr','',gene_pos_df$chromosome_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65a39937-d538-45ba-890c-5954e5a71ac4",
   "metadata": {},
   "source": [
    "# Run inferCNV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1ae8454a-9902-49b3-b908-1f5dec1cd5f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "runInferCNV = function(sce,sample){\n",
    "\n",
    "    # Create SPATA object\n",
    "    \n",
    "    spata_obj = initiateSpataObject_Empty(sample)\n",
    "    spata_obj@data[[sample]]$counts = Matrix::Matrix(sce@assays@data$counts[,sce$Sample==sample])\n",
    "    spata_obj@fdata[[sample]] = cbind(sce@colData[sce$Sample==sample,] %>% as.data.frame() %>% rownames_to_column('barcodes'),'sample'=sample)\n",
    "    spata_obj@coordinates[[sample]] = cbind(sce[,sce$Sample==sample]@int_colData$reducedDims$spatial,'sample'=sample)\n",
    "    \n",
    "    spata_obj <-\n",
    "        setDefaultInstructions(spata_obj) %>%\n",
    "        setDirectoryInstructions()\n",
    "\n",
    "    spata_obj <- setInitiationInfo(spata_obj)\n",
    "    \n",
    "    if (!dir.exists(paste0('./window100/',sample))){\n",
    "    dir.create(paste0('./window100/',sample),recursive = T)\n",
    "    }\n",
    "\n",
    "    # Run infercnv using SPATA2 functions\n",
    "    spata_obj <-\n",
    "      runCnvAnalysis_modified(\n",
    "        object = spata_obj,\n",
    "        directory_cnv_folder = paste0('./window100/',sample), # example directory\n",
    "        gene_pos_df = gene_pos_df,\n",
    "        cnv_prefix = \"Chr\",\n",
    "        ref_annotation = data.frame('barcodes' = colnames(P0_FT_1),'sample'='ref') %>% column_to_rownames('barcodes'), \n",
    "        ref_mtr = as.matrix(P0_FT_1@assays@data$counts), \n",
    "        ref_regions = cnv_ref$regions,\n",
    "        smooth_by_chromosome = list(window_length = 101, smooth_ends = TRUE),\n",
    "        define_signif_tumor_subclusters = list(p_val = 0.05, hclust_method = \"ward.D2\",\n",
    "            cluster_by_groups = FALSE, partition_method = \"qnorm\")\n",
    "        )\n",
    "    \n",
    "    confuns::give_feedback(msg = \"Trimming out ref in cnv matrix.\", verbose = T)\n",
    "    spata_obj@cnv[[sample]]$cnv_mtr = spata_obj@cnv[[sample]]$cnv_mtr[,colnames(spata_obj@cnv[[sample]]$cnv_mtr) %in% spata_obj@fdata[[sample]]$barcodes ]\n",
    "    \n",
    "    confuns::give_feedback(msg = \"Saving SPATA object.\", verbose = T)\n",
    "    saveRDS(spata_obj,paste0('./window100/',sample,'/spata_obj.RDS'))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9410adeb-5f47-40df-b0bc-89bd4ea6b592",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:18:56 Setting up new `spata2` object.\n",
      "\n",
      "Warning message in confuns::give_feedback(msg = msg, fdb.fn = fdb_fn, with.time = FALSE):\n",
      "“Argument `of_sample` is deprecated and no longer in use.( used by runCnvAnalysis_modified() )”\n",
      "14:18:57 Checking input for reference data.\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "“sparse->dense coercion: allocating vector of size 4.6 GiB”\n",
      "14:19:00 Combining input and reference data.\n",
      "\n",
      "14:19:04 Getting gene positions.\n",
      "\n",
      "14:19:04 Starting analysis pipeline.\n",
      "\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22m`invoke()` is deprecated as of rlang 0.4.0.\n",
      "Please use `exec()` or `inject()` instead.\n",
      "\u001b[90mThis warning is displayed once every 8 hours.\u001b[39m”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:19:05] ::order_reduce:Start.\n",
      "INFO [2025-11-21 14:19:05] .order_reduce(): expr and order match.\n",
      "INFO [2025-11-21 14:19:08] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 16147,125971 Total=191114865 Min=0 Max=277.\n",
      "INFO [2025-11-21 14:19:09] -filtering out cells < 100 or > Inf, removing 0.12622 % of cells\n",
      "WARN [2025-11-21 14:19:10] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
      "INFO [2025-11-21 14:19:13] validating infercnv_obj\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:19:14 Removing genes from matrix with mean expression below threshold.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:19:13] ::above_min_mean_expr_cutoff:Start\n",
      "INFO [2025-11-21 14:19:14] Removing 12625 genes from matrix as below mean expr threshold: 0.1\n",
      "INFO [2025-11-21 14:19:15] validating infercnv_obj\n",
      "INFO [2025-11-21 14:19:15] There are 3522 genes and 125812 cells remaining in the expr matrix.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:19:15 Removing low quality barcode spots.\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "“sparse->dense coercion: allocating vector of size 3.3 GiB”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:19:29] no genes removed due to min cells/gene filter\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:19:29 Normalizing counts by sequencing depth.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:19:29] normalizing counts matrix by depth\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in asMethod(object):\n",
      "“sparse->dense coercion: allocating vector of size 3.3 GiB”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:19:41] Computed total sum normalization factor as median libsize: 909.000000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:19:42 Conducting anscombe and logarithmic transformation.\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "“sparse->dense coercion: allocating vector of size 3.3 GiB”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:19:51] transforming log2xplus1()\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:20:01 Applying maximal threshold bounds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:20:01] ::process_data:setting max centered expr, threshold set to: +/-:  2.6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:20:12 Smoothing by chromosome.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:20:11] smooth_by_chromosome: chr: 1\n",
      "INFO [2025-11-21 14:21:46] smooth_by_chromosome: chr: 2\n",
      "INFO [2025-11-21 14:23:20] smooth_by_chromosome: chr: 3\n",
      "INFO [2025-11-21 14:24:43] smooth_by_chromosome: chr: 4\n",
      "INFO [2025-11-21 14:25:55] smooth_by_chromosome: chr: 5\n",
      "INFO [2025-11-21 14:27:10] smooth_by_chromosome: chr: 6\n",
      "INFO [2025-11-21 14:28:25] smooth_by_chromosome: chr: 7\n",
      "INFO [2025-11-21 14:29:45] smooth_by_chromosome: chr: 8\n",
      "INFO [2025-11-21 14:31:02] smooth_by_chromosome: chr: 9\n",
      "INFO [2025-11-21 14:32:21] smooth_by_chromosome: chr: 10\n",
      "INFO [2025-11-21 14:33:40] smooth_by_chromosome: chr: 11\n",
      "INFO [2025-11-21 14:35:02] smooth_by_chromosome: chr: 12\n",
      "INFO [2025-11-21 14:36:19] smooth_by_chromosome: chr: 13\n",
      "INFO [2025-11-21 14:37:00] smooth_by_chromosome: chr: 14\n",
      "INFO [2025-11-21 14:38:17] smooth_by_chromosome: chr: 15\n",
      "INFO [2025-11-21 14:39:31] smooth_by_chromosome: chr: 16\n",
      "INFO [2025-11-21 14:40:47] smooth_by_chromosome: chr: 17\n",
      "INFO [2025-11-21 14:42:02] smooth_by_chromosome: chr: 18\n",
      "INFO [2025-11-21 14:42:29] smooth_by_chromosome: chr: 19\n",
      "INFO [2025-11-21 14:43:52] smooth_by_chromosome: chr: 20\n",
      "INFO [2025-11-21 14:45:06] smooth_by_chromosome: chr: 21\n",
      "INFO [2025-11-21 14:45:36] smooth_by_chromosome: chr: 22\n",
      "INFO [2025-11-21 14:46:34] smooth_by_chromosome: chr: X\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:47:35 Centering cell expression across chromosome.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:47:35] ::center_smooth across chromosomes per cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:48:26 Subtracting reference expression from observed expression.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:48:25] ::subtract_ref_expr_from_obs:Start inv_log=TRUE, use_bounds=TRUE\n",
      "INFO [2025-11-21 14:48:25] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2025-11-21 14:48:39] -subtracting expr per gene, use_bounds=TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:49:00 Clearing noise via reference mean standard deviation.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:49:00] invert_log2(), computing 2^x\n",
      "INFO [2025-11-21 14:49:14] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
      "INFO [2025-11-21 14:49:18] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.966171457884711 - 1.03382854211529\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:49:27 Removing outliers.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2025-11-21 14:49:27] ::remove_outlier_norm:Start out_method: average_bound lower_bound: NA upper_bound: NA\n",
      "INFO [2025-11-21 14:49:27] ::remove_outlier_norm:Start out_method: average_bound lower_bound: NA upper_bound: NA\n",
      "INFO [2025-11-21 14:49:27] ::remove_outlier_norm using method: average_bound for defining outliers.\n",
      "INFO [2025-11-21 14:50:33] outlier bounds defined between: 0.942583 - 1.08848\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14:50:41 Copy number variation pipeline completed.\n",
      "\n",
      "14:50:41 Saving infercnv-object under './window100/P6_YF_1/infercnv-obj.RDS'.\n",
      "\n",
      "14:51:41 Summarizing cnv-results by chromosome.\n",
      "\n",
      "15:10:43 Adding results to the spata-object's feature data.\n",
      "\n",
      "Warning message in addFeatures(object = object, feature_df = ordered_cnv_df2, overwrite = TRUE):\n",
      "“Added features contain data for 125812 barcodes. Spata object contains 87959. Missing barcodes get NAs as values.”\n",
      "15:10:45 Infercnv Done.\n",
      "\n",
      "15:10:45 Trimming out ref in cnv matrix.\n",
      "\n",
      "15:10:46 Saving SPATA object.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "runInferCNV(P6_YF_1,'P6_YF_1')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "392afd4f-25e1-4bda-a9c9-3a44f780e9d7",
   "metadata": {},
   "source": [
    "# Convert to csv files to read in python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "93d7ed95-8dda-4050-8e40-8e0109d986d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "convert_spata = function(sample){\n",
    "    spata_obj = readRDS(paste0('./window100/',sample,'/spata_obj.RDS'))\n",
    "    write.csv(\n",
    "        spata_obj@cnv[[sample]]$cnv_mtr[,colnames(spata_obj@cnv[[sample]]$cnv_mtr) %in% spata_obj@fdata[[sample]]$barcodes],\n",
    "        paste0('./window100/',sample,'/cnv.csv')\n",
    "    )\n",
    "\n",
    "    metadata = spata_obj@fdata[[sample]] %>% column_to_rownames('barcodes') \n",
    "    metadata = metadata[colnames(spata_obj@cnv[[sample]]$cnv_mtr),]\n",
    "    write.csv(metadata,paste0('./window100/',sample,'/metadata.csv'))\n",
    "    \n",
    "    coord = spata_obj@coordinates[[sample]]\n",
    "    # rownames(coord) = coord$barcodes\n",
    "    coord = coord[colnames(spata_obj@cnv[[sample]]$cnv_mtr),c('x','y')]\n",
    "    write.csv(coord,paste0('./window100/',sample,'/coord.csv'))\n",
    "\n",
    "    \n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fdfb901d-445b-4a1a-afa2-c6dace21abaa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "convert_spata('P6_YF_1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ab49541-6f7e-4e66-a814-67bd706d3c03",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:conda-spata2]",
   "language": "R",
   "name": "conda-env-conda-spata2-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
